src/mergeDeep.ts

```diff
@@ -0,0 +1,117 @@
+import type { MergeDeep } from 'type-fest';
+import { purry } from './purry';
+
+type UnknownRecordOrArray =
+  | Record<PropertyKey, unknown>
+  | ReadonlyArray<unknown>;
+
+function isRecordOrArray(object: unknown): object is UnknownRecordOrArray {
+  return (
+    Array.isArray(object) ||
+    (typeof object === 'object' &&
+      object !== null &&
+      Object.getPrototypeOf(object) === Object.prototype)
+  );
+}
+
+/**
+ * Recursively merges two values, `a` and `b`.
+ * @param target - The first value to merge. This is the dominant parameter in the merge operation.
+ * @param source - The second value to merge.
+ * @signature
+ *    R.mergeDeep(a, b)
+ * @example
+ *    R.mergeDeep({ foo: 'bar', x: 1 }, { foo: 'baz', y: 2 }) // => { foo: 'bar', x: 1, y: 2 }
+ * @data_first
+ * @category Object
+ * @pipeable
+ */
+export function mergeDeep<
+  Target extends UnknownRecordOrArray,
+  Source extends UnknownRecordOrArray
+>(
+  target: Target,
+  source: Source
+): MergeDeep<Target, Source, { arrayMergeMode: 'spread' }>;
+
+/**
+ * Recursively merges two values, `a` and `b`.
+ * @param source - The second value to merge.
+ * @signature
+ *    R.mergeDeep(b)(a)
+ * @example
+ *    R.mergeDeep({ foo: 'baz', y: 2 })({ foo: 'bar', x: 1 }) // => { foo: 'bar', x: 1, y: 2 }
+ *    R.pipe(
+ *      { foo: 'bar', x: 1 },
+ *      R.mergeDeep({ foo: 'baz', y: 2 }),
+ *    ) // => { foo: 'bar', x: 1, y: 2 }
+ * @data_last
+ * @category Object
+ * @pipeable
+ */
+export function mergeDeep<
+  Target extends UnknownRecordOrArray,
+  Source extends UnknownRecordOrArray
+>(source: Source): (target: Target) => MergeDeep<Target, Source>;
+
+export function mergeDeep() {
+  return purry(_mergeDeep, arguments);
+}
+
+function _mergeDeep(a: UnknownRecordOrArray, b: UnknownRecordOrArray) {
+  const seenValues = new WeakSet(); // use to keep track of which objects have been seen.
+
+  function isArrayCyclic(arr: ReadonlyArray<unknown>) {
+    return arr.some(value => value && seenValues.has(value));
+  }
+
+  function merge(a: UnknownRecordOrArray, b: UnknownRecordOrArray) {
+    if (a === null || b === null) {
+      return a ?? b;
+    }
+
+    if (seenValues.has(a) || seenValues.has(b)) {
+      return a;
+    }
+    seenValues.add(a);
+    seenValues.add(b);
+
+    if (Array.isArray(a)) {
+      if (Array.isArray(b)) {
+        if (isArrayCyclic(a)) {
+          return b;
+        }
+        if (isArrayCyclic(b)) {
+          return a;
+        }
+        return a.concat(b);
+      }
+      return a;
```

Comment:
This is incorrect in regards to the type you defined, when there's a mismatch between one side being an array and the other being something else, the latter value is always used, never the former: https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgWQKZQOaoCKtWOAXzgDMoIQ4AiGATzFQFoTUBnGKgbgChu6G4ADTgBeFOiy58AHm4B6OXCUA9APzclSAIYAuOACMIEADaotAOyIAaDYji645gK4h96ANoBda7e1QoWrRomKjIEAAmqHpUrGBQZuFURNwAfDx89KhwAJqi4iFSYLIKSnBqvvZ67FDA5hhePpqVBkamFo12Wv6BwVhhkdGx8VqJyWm8-FkAWgCMeb04eEXyiirqTQ7VtfXehDYbes6uHrv7nd1BEqERUdRDCUmEqemTcFMATPNXhcWrZeudKowGp1Bp7CoOI5uKBgs5+AKXEL9W5UeJgYxaADGqEezwmmTeAGYvgUlr9SuUDnAtqDThDDi5oWC8UA

```suggestion
      return b;
```

Reply:
fixed